awx/main/tasks.py

```diff
@@ -3137,3 +3009,212 @@ def deep_copy_model_obj(
+        # we yank the socket out from underneath the processor, which will cause it
+        # to exit. A reference to the processor_future is passed into the cancel_watcher_future,
+        # Which exits if the job has finished normally. The context manager ensures we do not
+        # leave any threads laying around.
+        with concurrent.futures.ThreadPoolExecutor(max_workers=2) as executor:
+            processor_future = executor.submit(self.processor, resultfile)
+            cancel_watcher_future = executor.submit(self.cancel_watcher, processor_future)
+            futures = [processor_future, cancel_watcher_future]
+            first_future = concurrent.futures.wait(futures,
+                                                   return_when=concurrent.futures.FIRST_COMPLETED)
+
+            res = list(first_future.done)[0].result()
+            if res.status == 'canceled':
+                receptor_ctl.simple_command(f"work cancel {unit_id}")
+                resultsock.shutdown(socket.SHUT_RDWR)
+                resultfile.close()
+            elif res.status == 'error':
+                # TODO: There should be a more efficient way of getting this information
+                receptor_work_list = receptor_ctl.simple_command("work list")
+                raise RuntimeError(receptor_work_list[unit_id]['Detail'])
+
+        receptor_ctl.simple_command(f"work release {unit_id}")
+        return res
+
+    # Spawned in a thread so Receptor can start reading before we finish writing, we
+    # write our payload to the left side of our socketpair.
+    def transmit(self, _socket):
+        if self.work_type == 'local':
+            self.runner_params['only_transmit_kwargs'] = True
+
+        ansible_runner.interface.run(streamer='transmit',
+                                     _output=_socket.makefile('wb'),
+                                     **self.runner_params)
+
+        # Socket must be shutdown here, or the reader will hang forever.
+        _socket.shutdown(socket.SHUT_WR)
+
+    def processor(self, resultfile):
+        return ansible_runner.interface.run(streamer='process',
+                                            quiet=True,
+                                            _input=resultfile,
+                                            event_handler=self.task.event_handler,
+                                            finished_callback=self.task.finished_callback,
+                                            status_handler=self.task.status_handler,
+                                            **self.runner_params)
+
+    @property
+    def receptor_params(self):
+        if self.task.instance.is_container_group_task:
+            spec_yaml = yaml.dump(self.pod_definition, explicit_start=True)
+            kubeconfig_yaml = yaml.dump(self.kube_config, explicit_start=True)
+
+            receptor_params = {
+                "secret_kube_pod": spec_yaml,
+                "secret_kube_config": kubeconfig_yaml
```

Comment:
there are some cases we are seeing:

```
Traceback (most recent call last):
  File "/home/ec2-user/tower-qa/tests/api/external_jobs/test_container_groups.py", line 313, in test_container_group_sliced_jobs_exceeding_resource_quota
    job.assert_status(['running', 'pending', 'waiting'])
  File "/home/ec2-user/venvs/venv/lib64/python3.6/site-packages/awxkit/api/mixins/has_status.py", line 93, in assert_status
    raise AssertionError(msg)
AssertionError: Job-3243 has status of error, which is not in ['running', 'pending', 'waiting'].
result_traceback:
Traceback (most recent call last):
  File "/var/lib/awx/venv/awx/lib64/python3.6/site-packages/awx/main/tasks.py", line 1425, in run
    res = receptor_job.run()
  File "/var/lib/awx/venv/awx/lib64/python3.6/site-packages/awx/main/tasks.py", line 3036, in run
    params=self.receptor_params)
  File "/var/lib/awx/venv/awx/lib64/python3.6/site-packages/receptorctl/socket_interface.py", line 187, in submit_work
    raise RuntimeError(errmsg)
RuntimeError: Failed to start work unit: secret_kube_config provided but not allowed
```

Looks like 2 places? So not a lot, don't know what's up with that.

Reply:
raised from

https://github.com/project-receptor/receptor/blob/e293394235082df0b5f2cf037a43d13c6a1f3787/receptorctl/receptorctl/socket_interface.py#L187